File: Laba6/Laba6/Models/Task4Class.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Laba6.Models
{
    class PrintedPublication
    {
        protected string title;
        protected string author;
        protected int year;

        public PrintedPublication()
        {
            title = "Невідомий заголовок";
            author = "Невідомий автор";
            year = 2000;
        }

        public PrintedPublication(string title, string author, int year)
        {
            this.title = title;
            this.author = author;
            this.year = year;
        }

        public PrintedPublication(string title)
        {
            this.title = title;
            author = "Невідомий автор";
            year = 2000;
        }

        ~PrintedPublication()
        {
            Console.WriteLine($"Об'єкт {title} утилізовано.");
        }

        public virtual void DisplayInfo()
        {
            Console.WriteLine($"Назва: {title}");
            Console.WriteLine($"Автор: {author}");
            Console.WriteLine($"Рік видання: {year}");
        }

        public void Show()
        {
            Console.WriteLine("Інформація про об'єкт:");
            DisplayInfo();
        }
    }

    class Magazine : PrintedPublication
    {
        private string genre;

        public Magazine() : base()
        {
            genre = "Невідомий жанр";
        }

        public Magazine(string title, string author, int year, string genre) : base(title, author, year)
        {
            this.genre = genre;
        }

        public override void DisplayInfo()
        {
            base.DisplayInfo();
            Console.WriteLine($"Жанр: {genre}");
        }
    }

    class Book : PrintedPublication
    {
        private int pageCount;

        public Book() : base()
        {
            pageCount = 0;
        }

        public Book(string title, string author, int year, int pageCount) : base(title, author, year)
        {
            this.pageCount = pageCount;
        }

        public override void DisplayInfo()
        {
            base.DisplayInfo();
            Console.WriteLine($"Кількість сторінок: {pageCount}");
        }
    }

    class Textbook : Book
    {
        private string subject;

        public Textbook() : base()
        {
            subject = "Невідомий предмет";
        }

        public Textbook(string title, string author, int year, int pageCount, string subject) : base(title, author, year, pageCount)
        {
            this.subject = subject;
        }

        public override void DisplayInfo()
        {
            base.DisplayInfo();
            Console.WriteLine($"Предмет: {subject}");
        }
    }
}

File: Laba6-Python/TaskModels/task4Class.py
from PyQt5.QtWidgets import QApplication, QWidget, QVBoxLayout, QLabel, QPushButton

class PrintedPublication:
    def __init__(self, title="PrintedPublication", author="Невідомий автор", year=2000):
        self.title = title
        self.author = author
        self.year = year

    def get_price(self):
        return 30

    def show(self):
        return f"Назва: {self.title}\nАвтор: {self.author}\nРік видання: {self.year}"


class Magazine(PrintedPublication):
    def __init__(self, title="Magazine", author="Невідомий автор", year=2000, genre="Невідомий жанр"):
        super().__init__(title, author, year)
        self.genre = genre

    def get_price(self):
        return 50

    def show(self):
        return f"{super().show()}\nЖанр: {self.genre}"


class Book(PrintedPublication):
    def __init__(self, title="Book", author="Невідомий автор", year=2000, pageCount=0):
        super().__init__(title, author, year)
        self.page_count = pageCount

    def get_price(self):
        return 20

    def show(self):
        return f"{super().show()}\nКількість сторінок: {self.page_count}"


class Textbook(Book):
    def __init__(self, title="Textbook", author="Невідомий автор", year=2000, pageCount=0, subject="Невідомий предмет"):
        super().__init__(title, author, year, pageCount)
        self.subject = subject

    def get_price(self):
        return 70

    def show(self):
        return f"{super().show()}\nПредмет: {self.subject}"


class MainWindow(QWidget):
    def __init__(self):
        super().__init__()

        self.layout = QVBoxLayout()

        self.label = QLabel(self)
        self.layout.addWidget(self.label)

        self.button = QPushButton('Show Information', self)
        self.button.clicked.connect(self.show_information)
        self.layout.addWidget(self.button)

        self.setLayout(self.layout)

    def show_information(self):
        magazine = Magazine("National Geographic", "Various Authors", 2022, "Science")
        book = Book("The Great Gatsby", "F. Scott Fitzgerald", 1925, 180)
        textbook = Textbook("Mathematics for Beginners", "John Doe", 2020, 300, "Mathematics")

        information = f"\nMagazine:\n{magazine.show()}\nPrice: ${magazine.get_price()}\n\n" \
                      f"Book:\n{book.show()}\nPrice: ${book.get_price()}\n\n" \
                      f"Textbook:\n{textbook.show()}\nPrice: ${textbook.get_price()}"

        self.label.setText(information)